### DISCUSSION TITLE
How to contribute to OrchardCore?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#11825] [Q&A] [answered]
I am trying to help by contributing to OrchardCore. What is the correct workflow to contribute to the project?

Here is what I have been doing
 1. Forked the project [My Fork](https://github.com/CrestApps/OrchardCore)
 2. Before I want to work on a feature/bug fix I do the following
     1. On my main branch I fetch the main upstream (i.e., OrchardCMS:main) to sync my branch with the main project
        ![image](https://user-images.githubusercontent.com/24724371/172423914-ec76320c-cb47-4958-8436-c45352fba712.png)
     2. Now, that my branch has the latest code, I create a new feature branch based on my main branch (i.e., CreatApps:main)
     3. I make my code changed to my feature branch then commit my code.
 3. Now when the changes are done, I create a Pull Request based on my feature branch.
 4. However, I get too many commits being including in my Pull Request as you can see in this simple [PR](https://github.com/OrchardCMS/OrchardCore/pull/11824)

I am guessing the issue is that I "Fetch the main upstream" to sync my main branch with OrchardCMS:main.

I use GitHub's UI and VisualStudio 2022 to execute `git` commands.

Any tips are highly appreciated.

**Comment by Skrypt** [comment, marked as answer]
See the issue is that you main branch is desynchronized from the core project main branch.

![image](https://user-images.githubusercontent.com/3228637/172433890-9076b939-e880-4449-b4cd-addfd1c6e648.png)

You can see on the image there that it says that you are 22 commits ahead of the actual main branch of the base repository.
Something went wrong on your fork at some point which I can't really say. One thing that you can do is that you can try to delete the main branch entirely and to get it back locally on your PC and try to force push it on your fork.

**Reply by MikeAlhayek** [reply]
@Skrypt Thank you!
Not sure deleting is an option since I have multiple PR that are not yet merged. If I delete the fork, I wont be able to make any more updates. Right?

**Reply by ns8482e** [reply]
1. Fork 
2. Always rebase your fork's main branch from upstream
3. Always create new feature branch from your fork's rebased main to keep new feature current  
4. commit in your feature branch and create PR
5.  keep your feature branch/PR current with upstream's main by periodically pull/rebase from upstream's main.

**Reply by MikeAlhayek** [reply]
@Skrypt you identified the problem! I followed the steps [here](https://github.com/OrchardCMS/OrchardCore/discussions/11825#discussioncomment-2906356) to get a clean branch without losing my pull request. Once all these PR closed, I'll delete my fork and start over to eliminate confusion on my end.

**Comment by hishamco** [comment]
Simply use a Git client such as GitHub Client or Fork .. etc to make the process easy, some will fetch the latest commits automatically

**Comment by MikeAlhayek** [comment]
@hishamco @Skrypt I may have identified the issue. Trying to see if there is a way to first it.

![image](https://user-images.githubusercontent.com/24724371/172626068-d58742ee-b199-43e6-82aa-94c0ccb01a11.png)


It looks a while back I did something wrong where I created a PR in my own fork from main "not sure why or how :(". But is there a way to reset the main branch at that point in time and then somehow pull the latest all over again so my fork is back async with OrchardCMS:main?

![image](https://user-images.githubusercontent.com/24724371/172626499-fced6903-d5a0-4187-b0ba-19beb002e844.png)


I just want to make sure my feature branches are not impacted by this change

**Comment by hishamco** [comment]
When you start a new PR just right click the main branch and create yours, this way both are in sync

**Reply by MikeAlhayek** [reply]
I don't think I understand. If I create a new PR, it from a feature branch not the main branch. The issue seems to be because my main branch has commits that should not be there [here is a list of them](https://github.com/OrchardCMS/OrchardCore/compare/main...CrestApps:main) if I can somehow undo/reset/revert these commits, without losing my branches then I should be back a synced with OrchardCMS:main

**Comment by MikeAlhayek** [comment]
I created a new branch called `master` from a commit that was done before 11/13/2021. Then from there I fetch the latest to not mess with existing PR, yet to start from a good point. Moving forward, I would create Feature breach from master instead of main. I hope this fixes the issues with I submit a new PR

**Comment by hishamco** [comment]
What I see in #11829 is great, there's 1 commit instead, seems your issue has been resolved
